lib/streamlit/runtime/media_file_manager.py

```diff
@@ -231,3 +244,128 @@ def add(
+
+        with self._lock:
+            # Generate a unique placeholder ID for this deferred callable
+            file_id = uuid.uuid4().hex
+
+            # Store the callable with its metadata
+            self._deferred_callables[file_id] = {
+                "callable": data_callable,
+                "mimetype": mimetype,
+                "filename": file_name,
+                "coordinates": coordinates,
+            }
+
+            # Track this deferred file by session and coordinate
+            self._files_by_session_and_coord[session_id][coordinates] = file_id
+
+            return file_id
+
+    def execute_deferred(self, file_id: str) -> str:
+        """Execute a deferred callable and return the URL to the generated file.
+
+        This method retrieves the callable registered with add_deferred(),
+        executes it, stores the result, and returns a URL to access it.
+
+        Safe to call from any thread.
+
+        Parameters
+        ----------
+        file_id : str
+            The placeholder file_id returned by add_deferred().
+
+        Returns
+        -------
+        str
+            The URL that can be used to download the generated file.
+
+        Raises
+        ------
+        MediaFileStorageError
+            If the file_id is not found or if the callable execution fails.
+        """
+        # Retrieve deferred callable metadata while holding lock
+        with self._lock:
+            if file_id not in self._deferred_callables:
+                raise MediaFileStorageError(f"Deferred file {file_id} not found")
+
+            deferred = self._deferred_callables[file_id]
+
+        # Execute callable outside lock to avoid blocking other operations
+        try:
+            data = deferred["callable"]()
+        except Exception as e:
+            raise MediaFileStorageError(f"Callable execution failed: {e}") from e
+
+        # Convert data to bytes if needed
+        data_as_bytes: bytes
+        if isinstance(data, str):
+            data_as_bytes = data.encode()
+        elif isinstance(data, bytes):
+            data_as_bytes = data
+        elif isinstance(data, io.BytesIO):
+            data.seek(0)
+            data_as_bytes = data.getvalue()
+        elif isinstance(data, io.BufferedReader):
+            data.seek(0)
+            data_as_bytes = data.read()
+        elif isinstance(data, io.RawIOBase):
+            data.seek(0)
+            read_data = data.read()
+            data_as_bytes = read_data if read_data else b""
+        else:
+            raise MediaFileStorageError(
+                f"Callable returned unsupported type: {type(data)}"
+            )
+
+        # Store the generated data and get the actual file_id
+        with self._lock:
+            actual_file_id = self._storage.load_and_get_id(
+                data_as_bytes,
+                deferred["mimetype"],
+                MediaFileKind.DOWNLOADABLE,
+                deferred["filename"],
+            )
```

Comment:
I'm not sure if our current media file cleanups will be effective in this case: https://github.com/streamlit/streamlit/blob/3112973df8595900f2f59e0bad5f165900dd7a01/lib/streamlit/runtime/media_file_manager.py#L114
E.g., it might be that we are aggregating unused media files in the media file storage that never get cleaned up. This may require further consideration on how to properly handle this.